src/ProgressOnderwijsUtils.Analyzers/MustUseExpressionResultAnalyzer.cs

```diff
@@ -0,0 +1,72 @@
+using Microsoft.CodeAnalysis.Diagnostics;
+using System.Collections.Immutable;
+
+namespace ProgressOnderwijsUtils.Analyzers
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public sealed class MustUseExpressionResultAnalyzer : DiagnosticAnalyzer
+    {
+        public static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(
+            "POU1001",
+            "Result must be used",
+            "The result of the expression may not be ignored.",
+            "Functional",
+            DiagnosticSeverity.Warning,
+            true
+        );
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+            => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze);
+            context.EnableConcurrentExecution();
+            context.RegisterSyntaxNodeAction(Analyze, SyntaxKind.ExpressionStatement);
+        }
+
+        static void Analyze(SyntaxNodeAnalysisContext context)
+            => AnalyzeExpressionStatement(context, context.Node as ExpressionStatementSyntax);
+
+        static void AnalyzeExpressionStatement(SyntaxNodeAnalysisContext context, ExpressionStatementSyntax? syntax)
+        {
+            if (syntax == null || syntax.Expression.Kind() == SyntaxKind.SimpleAssignmentExpression) {
+                return;
+            }
+
+            var exprType = context.SemanticModel.GetTypeInfo(syntax.Expression).Type;
+            if (!ExpressionTypeCanBeIgnored(exprType)) {
+                context.ReportDiagnostic(Diagnostic.Create(Rule, syntax.GetLocation(), exprType));
+            }
+        }
+
+        static bool ExpressionTypeCanBeIgnored(ITypeSymbol? exprType)
+        {
+            if (exprType == null) {
+                return true;
+            } else if (exprType.TypeKind == TypeKind.Error) {
+                return true;
+            } else if (exprType.SpecialType == SpecialType.System_Void) {
+                return true;
+            } else if (IsUtilsUnit(exprType)) {
+                return true;
+            } else {
+                // TODO: first start wilt the maybe's, then apply to other/all types
+                return !(exprType.TypeKind == TypeKind.Struct && IsInUtilsCollectionsNamespace(exprType) && exprType.Name == "Maybe");
+            }
+        }
+
+        static bool IsUtilsUnit(ITypeSymbol exprType)
+            => exprType.TypeKind == TypeKind.Struct
+                && IsInUtilsCollectionsNamespace(exprType)
+                && exprType.Name == "Unit";
+
+        static bool IsInUtilsCollectionsNamespace(ITypeSymbol exprType)
+            => exprType.ContainingNamespace.Name == "Collections"
+                && exprType.ContainingNamespace.ContainingNamespace.Name == "ProgressOnderwijsUtils"
+                && exprType.ContainingNamespace.ContainingNamespace.ContainingNamespace.IsGlobalNamespace;

```

Comment:
Er is vast een manier om dit handiger te doen, maar het is altijd wel een puzzeltje met roslyn.

Zou kunnen dat je bijv. de string "ProgressOnderwijsUtils.Collections" als namespace kunt parsen en dan rechtstreeks die verglijken met ContainingNamespace; dergelijke dingen doet dacht ik ook de LowReward analyzer.

Reply:
Je bedoelt iets als:
```
       static string FormatNamespace(ISymbol symbol)
        {
            var stack = new Stack<string>();
            for (var ns = symbol.ContainingNamespace; !ns.IsGlobalNamespace; ns = ns.ContainingNamespace) {
                stack.Push(ns.Name);
            }

            return string.Join(".", stack);
        }
```

Reply:
Nee, ik had gehoopt dat er voor INamespaceSymbol iets is zoals NameSyntax.IsEquivalentTo oid, maar dat lijkt in de semantic model niet te bestaan voor zover ik kan googlen.

prima zo dus.

Wat wel een optie zou zijn (een slechtere, denk ik) is via een van de ToString-achtige methods de volledige naam verglijken, maar analyzers zijn al duur genoeg, en we willen onze eigen VS niet trager maken dan nodig, dus dit lijkt me beter.

Oftewel: brainstorm afgekeurd ;-).